File: hnaj-be/app/Exceptions/BookmarkException.php
<?php

namespace App\Exceptions;

use App\Enums\BookmarkErrorCode;
use RuntimeException;

/**
 * Lỗi nghiệp vụ của domain bookmark.
 * Mang sẵn HTTP status và mã lỗi để handler dựng error envelope chung.
 */
class BookmarkException extends RuntimeException
{
    public function __construct(
        public readonly BookmarkErrorCode $errorCode,
        string $message,
        public readonly int $status = 422,
    ) {
        parent::__construct($message);
    }
}

File: hnaj-be-java/src/main/java/com/hnaj/api/exception/BookmarkException.java
package com.hnaj.api.exception;

public class BookmarkException extends RuntimeException {

    private final BookmarkErrorCode errorCode;
    private final int status;

    public BookmarkException(BookmarkErrorCode errorCode, String message, int status) {
        super(message);
        this.errorCode = errorCode;
        this.status = status;
    }

    public BookmarkErrorCode getErrorCode() {
        return errorCode;
    }

    public int getStatus() {
        return status;
    }
}
